**Applied Data Science & AI**

---

| | |
|---|---|
| **Time allowed** | 1 hour 30 minutes |
| **Total marks** | 100 |

---

### The scenario

You have been handed the raw sales records of a laptop distributor operating across six
cities in Pakistan. The six CSV files were exported from different systems, at different
times, by different people. Nothing has been cleaned.

### The files

| File | What it holds |
|---|---|
| `transactions_2023.csv` | Orders placed during 2023 |
| `transactions_2024.csv` | Orders placed during 2024 - **exported from a different system** |
| `customers.csv` | The customer register |
| `products.csv` | The product catalogue |
| `regions.csv` | Region code lookup |
| `returns.csv` | Orders that came back |


### Rules

1. Work in pandas. Every answer must be code that runs.
2. Answer in order - later questions depend on DataFrames built in earlier ones.
3. Where a question asks you to explain something, write it as a comment in your code cell.

---


## Part A - Reading and Inspecting

*8 marks*

Nothing needs cleaning yet. Find out what you are holding.



### Question A1  
*4 marks - Medium*

Load all six CSV files and report the shape of each.

Then, for `transactions_2023`, state **which columns contain missing values and how many are missing in each**.

In [1]:
import pandas as pd
import re

In [2]:
customers = pd.read_csv("./customers.csv")
products = pd.read_csv("./products.csv")
regions = pd.read_csv("./regions.csv")
returns = pd.read_csv("./returns.csv")
transactions23 = pd.read_csv("./transactions_2023.csv")
transactions24 = pd.read_csv("./transactions_2024.csv")

In [3]:
customers.shape,products.shape,regions.shape,returns.shape,transactions23.shape,transactions24.shape

((262, 5), (64, 5), (10, 4), (540, 3), (1400, 10), (1700, 10))

In [4]:
transactions23.head(3)

,Order ID,Customer Code,Product ID,Qty,Sales Rep,Warehouse,Discount %,Order Month,Order Note,Source URL
0,ORD23-70001,CUST-PK-1221,PRD-871,17 units,Omar Siddiqui,WH-HYD-11,0%,Jun,[RUSH] deliver before friday,https://portal.techmazone.com.pk/orders/new
1,ORD23-70002,cust_pk_1239,PRD-864,3 units,NaN,WH/ISB/12,NaN,Jul,[URGENT] client waiting at site,http://b2b.dell-partners.pk/quote
2,ORD23-70003,Customer: PK-1178 (2022),PRD-864,19 units,NaN,wh-skt-12,7.5%,Aug,[HOLD] awaiting credit approval,https://shop.hpstore.pk/checkout


In [5]:
transactions23.columns

Index(['Order ID', 'Customer Code', 'Product ID', 'Qty', 'Sales Rep',
       'Warehouse', 'Discount %', 'Order Month', 'Order Note', 'Source URL'],
      dtype='object')

In [6]:
transactions23.isnull().sum()

Order ID            0
Customer Code       0
Product ID          0
Qty                 0
Sales Rep        1212
Warehouse           0
Discount %        144
Order Month         0
Order Note          0
Source URL          0
dtype: int64

- There are only Two columns **Sales Rep** & **Discount %** with missing values.
- Number of `Missing Values` in each Column: `Sales Rep : 1212, Discount % : 144`


### Question A2  
*4 marks - Medium*

Three short parts on `products` and `transactions_2023`, before any cleaning:

1. Return the first 5 rows and first 3 columns of `products` by **position**.
2. Return rows labelled 10 to 15 of `products` with the columns from `Product Details` to `Unit Price`, by **label**.
3. How many 2023 orders were placed at exactly a 15% discount?

In a comment, state how many rows part 2 returned and why it is not the same count as `iloc[10:15]`.

In [7]:
products.iloc[0:5,0:3]

,Product ID,Product Details,Category
0,PRD-507,Dell Latitude 5420 15.6inch,Laptop
1,PRD-514,Dell Latitude 7430 16.0inch,laptop
2,PRD-521,Dell Vostro 3510 17.3inch,LAPTOP
3,PRD-528,Dell Vostro 5620 13.3inch,Laptop
4,PRD-535,Dell Inspiron 3511 14.0inch,Ultrabook


In [8]:
products

,Product ID,Product Details,Category,Unit Price,Warranty (Months)
0,PRD-507,Dell Latitude 5420 15.6inch,Laptop,"PKR 325,000",36.0
1,PRD-514,Dell Latitude 7430 16.0inch,laptop,"PKR 364,000",12.0
2,PRD-521,Dell Vostro 3510 17.3inch,LAPTOP,"PKR 410,000",24.0
3,PRD-528,Dell Vostro 5620 13.3inch,Laptop,"PKR 72,000",12.0
4,PRD-535,Dell Inspiron 3511 14.0inch,Ultrabook,"PKR 85,000",12.0
...,...,...,...,...,...
59,PRD-920,Samsung Galaxy Book2 14.0inch,workstation,"PKR 298,000",24.0
60,PRD-927,Samsung Galaxy Book3 15.6inch,Laptop,"PKR 325,000",24.0
61,PRD-934,Samsung Notebook Flash 16.0inch,laptop,"PKR 364,000",24.0
62,PRD-941,Samsung Odyssey 15 17.3inch,LAPTOP,"PKR 410,000",24.0


In [9]:
# Return 6 rows including row 15 itself.
# df.loc[] uses label indexing instead of integer based indexing with df.iloc[]
# Because of label indexing and not integer indexing, the count is different with row 15[10:15] included.

products.loc[10:15, ["Product Details", "Category", "Unit Price"]]

,Product Details,Category,Unit Price
10,HP ProBook 640 15.6inch,Workstation,"PKR 189,000"
11,HP EliteBook 840 16.0inch,workstation,"PKR 214,000"
12,HP EliteBook 650 17.3inch,Laptop,"PKR 245,000"
13,HP Pavilion 15 13.3inch,laptop,"PKR 268,000"
14,HP Pavilion 14 14.0inch,LAPTOP,"PKR 298,000"
15,HP Victus 16 15.6inch,Laptop,"PKR 325,000"


In [10]:
transactions23[transactions23["Discount %"] == "15%"]

,Order ID,Customer Code,Product ID,Qty,Sales Rep,Warehouse,Discount %,Order Month,Order Note,Source URL
21,ORD23-70022,BR2021/PK-1117,PRD-759,17 units,NaN,WH/PEW/12,15%,Mar,send send updated quotation,https://www.samsung-b2b.pk/orders/create
23,ORD23-70024,2019-PK_1085,PRD-871,22 units,NaN,WH HYD 08,15%,May,very very late response from buyer,http://orders.lenovo-pk.com/renewal
40,ORD23-70041,Customer: PK-1083 (2017),PRD-836,39 units,NaN,WH-HYD-12,15%,Oct,customer requested extended warranty,https://www.acer-distributor.pk/cart
58,ORD23-70059,BR2016/PK-1012,PRD-619,25 units,NaN,wh-lhr-03,15%,Apr,credit terms extended to 60 days [approved],https://shop.hpstore.pk/enterprise
68,ORD23-70069,Customer: PK-1183 (2017),PRD-913,35 units,Hamza Sheikh,WH-LHR-09,15%,Feb,call the the customer before loading,http://reseller.msi-pakistan.pk/new
...,...,...,...,...,...,...,...,...,...,...
1354,ORD23-71355,2019-PK_1215,PRD-514,8 units,NaN,wh-rwp-02,15%,Apr,credit terms extended to 60 days [approved],https://shop.hpstore.pk/enterprise
1359,ORD23-71360,CUST-PK-1091,PRD-773,25 units,NaN,WH LHR 10,15%,Sep,[PENDING],http://orders.lenovo-pk.com/b2b/new
1360,ORD23-71361,2019-PK_1185,PRD-542,39 units,NaN,WH-KHI-04,15%,Oct,[CLOSED],https://www.acer-distributor.pk/cart
1376,ORD23-71377,cust_pk_1204,PRD-948,26 units,NaN,WH-HYD-02,15%,Feb,resend the E-Mail with pricing,http://reseller.msi-pakistan.pk/new


In [11]:
len(transactions23[transactions23["Discount %"] == "15%"])

144

---
## Part B - Cleaning Labels and Values

*18 marks*

Every column is text until you make it something else.



### Question B1  
*6 marks - Medium*

Every file has untidy column labels: trailing spaces, capitals, spaces between words, brackets, and a `%` sign.

Write **one function** that cleans a single label - strip the ends, drop `(` and `)`, turn `%` into `pct`, turn spaces into underscores, lowercase everything - and apply it to the labels of **all six** DataFrames.

In [12]:
# Basic function to apply string functions
def cleaning_labels(labels):
    labels = labels.strip()
    labels = labels.replace("(","")
    labels = labels.replace(")","")
    labels = labels.replace("%","pct")
    labels = labels.replace(" ","_")
    labels = labels.lower()

    return labels

#converting and saving all dataframes into list called 'list_columns_labels'.
list_columns_labels = [customers,products, regions, returns,transactions23, transactions24]

#using for loop and applying to each label
for label in list_columns_labels:
    label.columns = [cleaning_labels(col) for col in label.columns]

In [13]:
transactions23

,order_id,customer_code,product_id,qty,sales_rep,warehouse,discount_pct,order_month,order_note,source_url
0,ORD23-70001,CUST-PK-1221,PRD-871,17 units,Omar Siddiqui,WH-HYD-11,0%,Jun,[RUSH] deliver before friday,https://portal.techmazone.com.pk/orders/new
1,ORD23-70002,cust_pk_1239,PRD-864,3 units,NaN,WH/ISB/12,NaN,Jul,[URGENT] client waiting at site,http://b2b.dell-partners.pk/quote
2,ORD23-70003,Customer: PK-1178 (2022),PRD-864,19 units,NaN,wh-skt-12,7.5%,Aug,[HOLD] awaiting credit approval,https://shop.hpstore.pk/checkout
3,ORD23-70004,2019-PK_1035,PRD-885,12 units,NaN,WH SKT 08,10%,Sep,[PRIORITY] escalated by manager,http://orders.lenovo-pk.com/b2b/new
4,ORD23-70005,CUST-PK-1216,PRD-794,14 units,NaN,WH-FSD-03,7.5%,Oct,[BACKORDER] awaiting stock from vendor,https://www.acer-distributor.pk/cart
...,...,...,...,...,...,...,...,...,...,...
1395,ORD23-71396,BR2016/PK-1202,PRD-682,9 units,Omar Siddiqui,WH RWP 10,NaN,Sep,[PRIORITY] escalated by manager,http://orders.lenovo-pk.com/b2b/new
1396,ORD23-71397,2019-PK_1145,PRD-514,21 units,NaN,WH-HYD-06,NaN,Oct,[BACKORDER] awaiting stock from vendor,https://www.acer-distributor.pk/cart
1397,ORD23-71398,Customer: PK-1178 (2022),PRD-626,34 units,NaN,WH/SKT/08,12.5%,Nov,[SPLIT] shipping in two lots,https://portal.techmazone.com.pk/orders/repeat
1398,ORD23-71399,BR2016/PK-1162,PRD-920,2 units,NaN,wh-fsd-11,5%,Dec,confirm stock before dispatch [pending],http://b2b.dell-partners.pk/renewal



### Question B2  
*5 marks - Medium*

In **both** transaction files convert the quantity column (`18 units`) to an **integer** and the discount column (`12.5%`, some blank) to a **float** where blank means **0**.

The 2024 export does not use the same column labels as 2023. Do not rename anything yet.

In [14]:
# for quantity columns:
transactions23["qty"] = transactions23["qty"].str.strip('units').astype(int)
transactions24["quantity"] = transactions24["quantity"].str.strip('units').astype(int)

In [15]:
# for discount columns
transactions23["discount_pct"] = transactions23["discount_pct"].str.strip("%").astype(float)
transactions24["discount"] = transactions24["discount"].str.strip("%").astype(float)

In [16]:
print(transactions23.info(),transactions24.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1400 entries, 0 to 1399
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   order_id       1400 non-null   object 
 1   customer_code  1400 non-null   object 
 2   product_id     1400 non-null   object 
 3   qty            1400 non-null   int64  
 4   sales_rep      188 non-null    object 
 5   warehouse      1400 non-null   object 
 6   discount_pct   1256 non-null   float64
 7   order_month    1400 non-null   object 
 8   order_note     1400 non-null   object 
 9   source_url     1400 non-null   object 
dtypes: float64(1), int64(1), object(8)
memory usage: 109.5+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1700 entries, 0 to 1699
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   order_no        1700 non-null   object 
 1   customer_id     1700 non-null   object 
 2   item_code



### Question B3  
*7 marks - Hard*

In `products`:

1. Turn `PKR 325,000` into the integer `325000`.
2. Fill missing warranty values with `12` and make the column an integer.
3. **Without regex**, create `brand` (the first word of `product_details`) and `screen_size_inches` (the trailing size, as a float).
4. `category` in `products` and `segment_type` in `customers` are both inconsistent - mixed capitals and stray spaces, and `segment_type` has blanks. Standardise both; blank segments should read `unknown`. Report the value counts and state how many genuinely distinct values exist in each.

In [17]:
products

,product_id,product_details,category,unit_price,warranty_months
0,PRD-507,Dell Latitude 5420 15.6inch,Laptop,"PKR 325,000",36.0
1,PRD-514,Dell Latitude 7430 16.0inch,laptop,"PKR 364,000",12.0
2,PRD-521,Dell Vostro 3510 17.3inch,LAPTOP,"PKR 410,000",24.0
3,PRD-528,Dell Vostro 5620 13.3inch,Laptop,"PKR 72,000",12.0
4,PRD-535,Dell Inspiron 3511 14.0inch,Ultrabook,"PKR 85,000",12.0
...,...,...,...,...,...
59,PRD-920,Samsung Galaxy Book2 14.0inch,workstation,"PKR 298,000",24.0
60,PRD-927,Samsung Galaxy Book3 15.6inch,Laptop,"PKR 325,000",24.0
61,PRD-934,Samsung Notebook Flash 16.0inch,laptop,"PKR 364,000",24.0
62,PRD-941,Samsung Odyssey 15 17.3inch,LAPTOP,"PKR 410,000",24.0


In [18]:
products["unit_price"] = (products["unit_price"].str.replace("PKR","").str.replace(",","").str.strip().astype(int))
products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64 entries, 0 to 63
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   product_id       64 non-null     object 
 1   product_details  64 non-null     object 
 2   category         64 non-null     object 
 3   unit_price       64 non-null     int64  
 4   warranty_months  58 non-null     float64
dtypes: float64(1), int64(1), object(3)
memory usage: 2.6+ KB


In [19]:
products["warranty_months"] = products["warranty_months"].fillna(12).astype(int)
products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64 entries, 0 to 63
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   product_id       64 non-null     object
 1   product_details  64 non-null     object
 2   category         64 non-null     object
 3   unit_price       64 non-null     int64 
 4   warranty_months  64 non-null     int64 
dtypes: int64(2), object(3)
memory usage: 2.6+ KB


In [20]:
products["product_details"].value_counts()

product_details
Dell Latitude 5420 15.6inch      1
Dell Latitude 7430 16.0inch      1
Acer Predator 300 14.0inch       1
Acer Chromebook 314 15.6inch     1
Asus ExpertBook B1 16.0inch      1
                                ..
Lenovo Yoga 7 16.0inch           1
Acer Aspire 5 17.3inch           1
Acer Aspire 7 13.3inch           1
Acer Swift 3 14.0inch            1
Samsung Galaxy Book4 13.3inch    1
Name: count, Length: 64, dtype: int64

In [21]:
brand_split = products["product_details"].str.split(n=3,expand=True)
brand_split

,0,1,2,3
0,Dell,Latitude,5420,15.6inch
1,Dell,Latitude,7430,16.0inch
2,Dell,Vostro,3510,17.3inch
3,Dell,Vostro,5620,13.3inch
4,Dell,Inspiron,3511,14.0inch
...,...,...,...,...
59,Samsung,Galaxy,Book2,14.0inch
60,Samsung,Galaxy,Book3,15.6inch
61,Samsung,Notebook,Flash,16.0inch
62,Samsung,Odyssey,15,17.3inch


In [22]:
# saving the brand split in particular columns:

products["brand"] = brand_split[0]
products["screen_size_inches"] = brand_split[3]
products

,product_id,product_details,category,unit_price,warranty_months,brand,screen_size_inches
0,PRD-507,Dell Latitude 5420 15.6inch,Laptop,325000,36,Dell,15.6inch
1,PRD-514,Dell Latitude 7430 16.0inch,laptop,364000,12,Dell,16.0inch
2,PRD-521,Dell Vostro 3510 17.3inch,LAPTOP,410000,24,Dell,17.3inch
3,PRD-528,Dell Vostro 5620 13.3inch,Laptop,72000,12,Dell,13.3inch
4,PRD-535,Dell Inspiron 3511 14.0inch,Ultrabook,85000,12,Dell,14.0inch
...,...,...,...,...,...,...,...
59,PRD-920,Samsung Galaxy Book2 14.0inch,workstation,298000,24,Samsung,14.0inch
60,PRD-927,Samsung Galaxy Book3 15.6inch,Laptop,325000,24,Samsung,15.6inch
61,PRD-934,Samsung Notebook Flash 16.0inch,laptop,364000,24,Samsung,16.0inch
62,PRD-941,Samsung Odyssey 15 17.3inch,LAPTOP,410000,24,Samsung,17.3inch


In [23]:
products["category"] = products["category"].str.strip().str.title()
products

,product_id,product_details,category,unit_price,warranty_months,brand,screen_size_inches
0,PRD-507,Dell Latitude 5420 15.6inch,Laptop,325000,36,Dell,15.6inch
1,PRD-514,Dell Latitude 7430 16.0inch,Laptop,364000,12,Dell,16.0inch
2,PRD-521,Dell Vostro 3510 17.3inch,Laptop,410000,24,Dell,17.3inch
3,PRD-528,Dell Vostro 5620 13.3inch,Laptop,72000,12,Dell,13.3inch
4,PRD-535,Dell Inspiron 3511 14.0inch,Ultrabook,85000,12,Dell,14.0inch
...,...,...,...,...,...,...,...
59,PRD-920,Samsung Galaxy Book2 14.0inch,Workstation,298000,24,Samsung,14.0inch
60,PRD-927,Samsung Galaxy Book3 15.6inch,Laptop,325000,24,Samsung,15.6inch
61,PRD-934,Samsung Notebook Flash 16.0inch,Laptop,364000,24,Samsung,16.0inch
62,PRD-941,Samsung Odyssey 15 17.3inch,Laptop,410000,24,Samsung,17.3inch


In [24]:
customers["segment_type"] = (customers["segment_type"].str.strip().str.title().fillna("Unknown").replace("","Unknown"))
customers

,customer_code,customer_name,segment_type,home_city,signup_year
0,CUST-PK-1001,Yasir Distributors,Retail,Karachi,2015
1,BR2016/PK-1002,Laiba Systems,Retail,Hyderabad,2016
2,Customer: PK-1003 (2017),Iqra Traders,Retail,Lahore,2017
3,cust_pk_1004,Yusra Electronics,Retail,Faisalabad,2018
4,2019-PK_1005,Fahad Traders,Corporate,Multan,2019
...,...,...,...,...,...
257,Customer: PK-1258 (2022),Imran Electronics,Corporate,Rawalpindi,2022
258,cust_pk_1259,Saad & Sons,Corporate,Peshawar,2023
259,2024-PK_1260,Jawad Associates,Corporate,Quetta,2024
260,CUST-PK-1261,Fahad Systems,Government,Karachi,2015


In [25]:
print("Product Category:")
print(products["category"].value_counts())
print(f"Distinct Categories: {products["category"].nunique()}")

Product Category:
category
Laptop         24
Ultrabook      15
Gaming         15
Workstation    10
Name: count, dtype: int64
Distinct Categories: 4


In [26]:
print("segment_type:")
print(customers["segment_type"].value_counts())
print(f"Distinct Categories: {customers["segment_type"].nunique()}")

segment_type:
segment_type
Retail        60
Corporate     57
Government    54
Education     42
Reseller      42
Unknown        7
Name: count, dtype: int64
Distinct Categories: 6


---
## Part C - Filling the Gaps with Logic

*13 marks*



### Question C1  
*9 marks - Hard*

The sales rep is written **once** at the top of a block of orders; the rows underneath are blank and belong to that same rep, until the next name appears.

Fill in every blank in `transactions_2023`.

Then report how many orders each rep handled, and confirm no blanks remain.

In [27]:
transactions23["sales_rep"].value_counts()

sales_rep
Fatima Zafar     28
Omar Siddiqui    27
Sarah Nadeem     22
Ali Raza         21
Bilal Ansari     21
Hamza Sheikh     21
Hira Baig        18
Ayesha Malik     18
Salman Tariq     12
Name: count, dtype: int64

In [28]:
transactions23["sales_rep"].isnull().sum()

np.int64(1212)

In [29]:
transactions23["sales_rep"] = transactions23["sales_rep"].ffill()

In [30]:
transactions23["sales_rep"].value_counts()

sales_rep
Omar Siddiqui    208
Fatima Zafar     195
Ali Raza         183
Bilal Ansari     161
Sarah Nadeem     154
Hamza Sheikh     141
Ayesha Malik     138
Hira Baig        128
Salman Tariq      92
Name: count, dtype: int64

In [31]:
transactions23["sales_rep"].isnull().sum()

np.int64(0)


### Question C2  
*4 marks - Hard*

Apply the same fill to `transactions_2024` - remember it labels the column differently.

Then answer in a comment: if you had **stacked the two years first** and filled afterwards, what could go wrong? Be specific about the mechanism.

In [32]:
transactions24["representative"].value_counts()

representative
Sarah Nadeem     32
Nimra Qureshi    24
Hira Baig        21
Salman Tariq     21
Bilal Ansari     20
Fatima Zafar     20
Ayesha Malik     19
Adnan Hashmi     18
Hamza Sheikh     17
Omar Siddiqui    17
Ali Raza         16
Name: count, dtype: int64

In [33]:
transactions24["representative"] = transactions24["representative"].ffill()
transactions24

,order_no,customer_id,item_code,quantity,representative,warehouse_code,discount,month,note,source_link
0,ORD24-90001,Customer: PK-1048 (2022),PRD-570,7,Adnan Hashmi,WH-SKT-02,15.0,Feb,[RUSH] deliver before friday,https://portal.techmazone.com.pk/orders/new
1,ORD24-90002,BR2016/PK-1022,PRD-871,26,Adnan Hashmi,WH/SKT/12,0.0,Mar,[URGENT] client waiting at site,http://b2b.dell-partners.pk/quote
2,ORD24-90003,CUST-PK-1071,PRD-836,23,Adnan Hashmi,wh-qta-01,15.0,Apr,[HOLD] awaiting credit approval,https://shop.hpstore.pk/checkout
3,ORD24-90004,2019-PK_1145,PRD-759,21,Adnan Hashmi,WH MUL 12,5.0,May,[PRIORITY] escalated by manager,http://orders.lenovo-pk.com/b2b/new
4,ORD24-90005,CUST-PK-1086,PRD-892,13,Adnan Hashmi,WH-FSD-07,NaN,Jun,[BACKORDER] awaiting stock from vendor,https://www.acer-distributor.pk/cart
...,...,...,...,...,...,...,...,...,...,...
1695,ORD24-91690,BR2016/PK-1072,PRD-528,7,Hamza Sheikh,WH/PEW/07,5.0,Nov,installation scheduled next week [planned],https://www.samsung-b2b.pk/orders/create
1696,ORD24-91691,Customer: PK-1078 (2022),PRD-738,40,Hamza Sheikh,wh-qta-03,10.0,Dec,credit terms extended to 60 days [approved],https://shop.hpstore.pk/enterprise
1697,ORD24-91692,2019-PK_1005,PRD-626,12,Hamza Sheikh,WH LHR 12,5.0,Jan,serial numbers shared with buyer [done],http://orders.lenovo-pk.com/renewal
1698,ORD24-91693,cust_pk_1004,PRD-668,7,Hamza Sheikh,WH-ISB-03,7.5,Feb,[VERIFIED],https://portal.techmazone.com.pk/orders/new


In [34]:
transactions23["sales_rep"].isnull().sum()

np.int64(0)

- if you had stacked the transactions23 and transactions24 years first and filled afterwards, the 2nd year would be filled with first years last name
- Example: `transaction_2023's sales_rep('Sarah Nadeem')` would become `transactions24's` first representative

---
## Part D - Regular Expressions

*36 marks*

The largest section. The same information has been recorded five different ways by five different systems. Every question here has a wrong answer that runs without error - check your patterns against the data, not against your expectations.



### Question D1  
*7 marks - Hard*

The customer code has been entered five different ways across the files:

```
CUST-PK-1001        BR2019/PK-1002        Customer: PK-1003 (2020)
cust_pk_1004        2022-PK_1005
```

Some of these layouts also contain a **four-digit year**. A pattern of `(\d{4})` will therefore extract the wrong number from part of the file.

1. Write **one** regular expression that extracts the correct customer number from **all five** layouts into an integer column `cust_id`. Apply it to `transactions_2023`, `transactions_2024` and `customers`.
2. Prove your pattern was necessary: report **how many rows a naive `(\d{4})` would have got wrong**.

In [35]:
pattern = r"[Pp][Kk][-_ ](\d{4})"
customers["cust_id"] = customers["customer_code"].str.extract(pattern).astype(int)
customers["customer_code"]

0                  CUST-PK-1001
1                BR2016/PK-1002
2      Customer: PK-1003 (2017)
3                  cust_pk_1004
4                  2019-PK_1005
                 ...           
257    Customer: PK-1258 (2022)
258                cust_pk_1259
259                2024-PK_1260
260                CUST-PK-1261
261              BR2016/PK-1262
Name: customer_code, Length: 262, dtype: object

In [36]:
customers["customer_code"]

0                  CUST-PK-1001
1                BR2016/PK-1002
2      Customer: PK-1003 (2017)
3                  cust_pk_1004
4                  2019-PK_1005
                 ...           
257    Customer: PK-1258 (2022)
258                cust_pk_1259
259                2024-PK_1260
260                CUST-PK-1261
261              BR2016/PK-1262
Name: customer_code, Length: 262, dtype: object

In [37]:
pattern = r"[Pp][Kk][-_ ](\d{4})"
transactions23["cust_id"] = transactions23["customer_code"].str.extract(pattern).astype(int)
transactions23

,order_id,customer_code,product_id,qty,sales_rep,warehouse,discount_pct,order_month,order_note,source_url,cust_id
0,ORD23-70001,CUST-PK-1221,PRD-871,17,Omar Siddiqui,WH-HYD-11,0.0,Jun,[RUSH] deliver before friday,https://portal.techmazone.com.pk/orders/new,1221
1,ORD23-70002,cust_pk_1239,PRD-864,3,Omar Siddiqui,WH/ISB/12,NaN,Jul,[URGENT] client waiting at site,http://b2b.dell-partners.pk/quote,1239
2,ORD23-70003,Customer: PK-1178 (2022),PRD-864,19,Omar Siddiqui,wh-skt-12,7.5,Aug,[HOLD] awaiting credit approval,https://shop.hpstore.pk/checkout,1178
3,ORD23-70004,2019-PK_1035,PRD-885,12,Omar Siddiqui,WH SKT 08,10.0,Sep,[PRIORITY] escalated by manager,http://orders.lenovo-pk.com/b2b/new,1035
4,ORD23-70005,CUST-PK-1216,PRD-794,14,Omar Siddiqui,WH-FSD-03,7.5,Oct,[BACKORDER] awaiting stock from vendor,https://www.acer-distributor.pk/cart,1216
...,...,...,...,...,...,...,...,...,...,...,...
1395,ORD23-71396,BR2016/PK-1202,PRD-682,9,Omar Siddiqui,WH RWP 10,NaN,Sep,[PRIORITY] escalated by manager,http://orders.lenovo-pk.com/b2b/new,1202
1396,ORD23-71397,2019-PK_1145,PRD-514,21,Omar Siddiqui,WH-HYD-06,NaN,Oct,[BACKORDER] awaiting stock from vendor,https://www.acer-distributor.pk/cart,1145
1397,ORD23-71398,Customer: PK-1178 (2022),PRD-626,34,Omar Siddiqui,WH/SKT/08,12.5,Nov,[SPLIT] shipping in two lots,https://portal.techmazone.com.pk/orders/repeat,1178
1398,ORD23-71399,BR2016/PK-1162,PRD-920,2,Omar Siddiqui,wh-fsd-11,5.0,Dec,confirm stock before dispatch [pending],http://b2b.dell-partners.pk/renewal,1162


In [38]:
pattern = r"[Pp][Kk][-_ ](\d{4})"
transactions24["cust_id"] = transactions24["customer_id"].str.extract(pattern).astype(int)
transactions24


,order_no,customer_id,item_code,quantity,representative,warehouse_code,discount,month,note,source_link,cust_id
0,ORD24-90001,Customer: PK-1048 (2022),PRD-570,7,Adnan Hashmi,WH-SKT-02,15.0,Feb,[RUSH] deliver before friday,https://portal.techmazone.com.pk/orders/new,1048
1,ORD24-90002,BR2016/PK-1022,PRD-871,26,Adnan Hashmi,WH/SKT/12,0.0,Mar,[URGENT] client waiting at site,http://b2b.dell-partners.pk/quote,1022
2,ORD24-90003,CUST-PK-1071,PRD-836,23,Adnan Hashmi,wh-qta-01,15.0,Apr,[HOLD] awaiting credit approval,https://shop.hpstore.pk/checkout,1071
3,ORD24-90004,2019-PK_1145,PRD-759,21,Adnan Hashmi,WH MUL 12,5.0,May,[PRIORITY] escalated by manager,http://orders.lenovo-pk.com/b2b/new,1145
4,ORD24-90005,CUST-PK-1086,PRD-892,13,Adnan Hashmi,WH-FSD-07,NaN,Jun,[BACKORDER] awaiting stock from vendor,https://www.acer-distributor.pk/cart,1086
...,...,...,...,...,...,...,...,...,...,...,...
1695,ORD24-91690,BR2016/PK-1072,PRD-528,7,Hamza Sheikh,WH/PEW/07,5.0,Nov,installation scheduled next week [planned],https://www.samsung-b2b.pk/orders/create,1072
1696,ORD24-91691,Customer: PK-1078 (2022),PRD-738,40,Hamza Sheikh,wh-qta-03,10.0,Dec,credit terms extended to 60 days [approved],https://shop.hpstore.pk/enterprise,1078
1697,ORD24-91692,2019-PK_1005,PRD-626,12,Hamza Sheikh,WH LHR 12,5.0,Jan,serial numbers shared with buyer [done],http://orders.lenovo-pk.com/renewal,1005
1698,ORD24-91693,cust_pk_1004,PRD-668,7,Hamza Sheikh,WH-ISB-03,7.5,Feb,[VERIFIED],https://portal.techmazone.com.pk/orders/new,1004


In [39]:
# Naive pattern:
naive_pattern = customers["customer_code"].str.extract(r"(\d{4})").astype(int)
naive_pattern

,0
0,1001
1,2016
2,1003
3,1004
4,2019
...,...
257,1258
258,1259
259,2024
260,1261


In [40]:
naive_pattern = customers["customer_code"].str.extract(r"(\d{4})").astype(int)
not_naive_pattern = customers["customer_code"].str.extract(r"[Pp][Kk][-_](\d{4})").astype(int)
# print(not_naive_pattern)
# print(naive_pattern)

difference = (naive_pattern[0] != not_naive_pattern[0])

print("Difference Between Naive and Non-Naive Pattern")
print(pd.concat([naive_pattern[difference],
                 not_naive_pattern[difference]], axis=1))


Difference Between Naive and Non-Naive Pattern
        0     0
1    2016  1002
4    2019  1005
6    2021  1007
9    2024  1010
11   2016  1012
..    ...   ...
251  2016  1252
254  2019  1255
256  2021  1257
259  2024  1260
261  2016  1262

[105 rows x 2 columns]



### Question D2  
*5 marks - Hard*

Warehouse codes have been entered four ways:

```
WH-LHR-09        WH/ISB/04        wh-isb-05        WH ISB 08
```

1. Write **one** regular expression that pulls out the three-letter region code from all four, and store it as **uppercase** in a column `region_code` on both transaction files.
2. Report how many rows a stricter pattern of `WH-([A-Z]{3})-` would have **silently failed** to match.

In [41]:
pattern = r"WH[-/ ]([\w+]{3})[-/ ]"
transactions23["region_code"] = (transactions23["warehouse"].str.extract(pattern, expand=False).str.upper())
transactions23

,order_id,customer_code,product_id,qty,sales_rep,warehouse,discount_pct,order_month,order_note,source_url,cust_id,region_code
0,ORD23-70001,CUST-PK-1221,PRD-871,17,Omar Siddiqui,WH-HYD-11,0.0,Jun,[RUSH] deliver before friday,https://portal.techmazone.com.pk/orders/new,1221,HYD
1,ORD23-70002,cust_pk_1239,PRD-864,3,Omar Siddiqui,WH/ISB/12,NaN,Jul,[URGENT] client waiting at site,http://b2b.dell-partners.pk/quote,1239,ISB
2,ORD23-70003,Customer: PK-1178 (2022),PRD-864,19,Omar Siddiqui,wh-skt-12,7.5,Aug,[HOLD] awaiting credit approval,https://shop.hpstore.pk/checkout,1178,NaN
3,ORD23-70004,2019-PK_1035,PRD-885,12,Omar Siddiqui,WH SKT 08,10.0,Sep,[PRIORITY] escalated by manager,http://orders.lenovo-pk.com/b2b/new,1035,SKT
4,ORD23-70005,CUST-PK-1216,PRD-794,14,Omar Siddiqui,WH-FSD-03,7.5,Oct,[BACKORDER] awaiting stock from vendor,https://www.acer-distributor.pk/cart,1216,FSD
...,...,...,...,...,...,...,...,...,...,...,...,...
1395,ORD23-71396,BR2016/PK-1202,PRD-682,9,Omar Siddiqui,WH RWP 10,NaN,Sep,[PRIORITY] escalated by manager,http://orders.lenovo-pk.com/b2b/new,1202,RWP
1396,ORD23-71397,2019-PK_1145,PRD-514,21,Omar Siddiqui,WH-HYD-06,NaN,Oct,[BACKORDER] awaiting stock from vendor,https://www.acer-distributor.pk/cart,1145,HYD
1397,ORD23-71398,Customer: PK-1178 (2022),PRD-626,34,Omar Siddiqui,WH/SKT/08,12.5,Nov,[SPLIT] shipping in two lots,https://portal.techmazone.com.pk/orders/repeat,1178,SKT
1398,ORD23-71399,BR2016/PK-1162,PRD-920,2,Omar Siddiqui,wh-fsd-11,5.0,Dec,confirm stock before dispatch [pending],http://b2b.dell-partners.pk/renewal,1162,NaN


In [42]:
pattern = r"WH[-/ ]([\w+]{3})[-/ ]"
transactions24["region_code"] = (transactions24["warehouse_code"].str.extract(pattern, expand=False).str.upper())
transactions24

,order_no,customer_id,item_code,quantity,representative,warehouse_code,discount,month,note,source_link,cust_id,region_code
0,ORD24-90001,Customer: PK-1048 (2022),PRD-570,7,Adnan Hashmi,WH-SKT-02,15.0,Feb,[RUSH] deliver before friday,https://portal.techmazone.com.pk/orders/new,1048,SKT
1,ORD24-90002,BR2016/PK-1022,PRD-871,26,Adnan Hashmi,WH/SKT/12,0.0,Mar,[URGENT] client waiting at site,http://b2b.dell-partners.pk/quote,1022,SKT
2,ORD24-90003,CUST-PK-1071,PRD-836,23,Adnan Hashmi,wh-qta-01,15.0,Apr,[HOLD] awaiting credit approval,https://shop.hpstore.pk/checkout,1071,NaN
3,ORD24-90004,2019-PK_1145,PRD-759,21,Adnan Hashmi,WH MUL 12,5.0,May,[PRIORITY] escalated by manager,http://orders.lenovo-pk.com/b2b/new,1145,MUL
4,ORD24-90005,CUST-PK-1086,PRD-892,13,Adnan Hashmi,WH-FSD-07,NaN,Jun,[BACKORDER] awaiting stock from vendor,https://www.acer-distributor.pk/cart,1086,FSD
...,...,...,...,...,...,...,...,...,...,...,...,...
1695,ORD24-91690,BR2016/PK-1072,PRD-528,7,Hamza Sheikh,WH/PEW/07,5.0,Nov,installation scheduled next week [planned],https://www.samsung-b2b.pk/orders/create,1072,PEW
1696,ORD24-91691,Customer: PK-1078 (2022),PRD-738,40,Hamza Sheikh,wh-qta-03,10.0,Dec,credit terms extended to 60 days [approved],https://shop.hpstore.pk/enterprise,1078,NaN
1697,ORD24-91692,2019-PK_1005,PRD-626,12,Hamza Sheikh,WH LHR 12,5.0,Jan,serial numbers shared with buyer [done],http://orders.lenovo-pk.com/renewal,1005,LHR
1698,ORD24-91693,cust_pk_1004,PRD-668,7,Hamza Sheikh,WH-ISB-03,7.5,Feb,[VERIFIED],https://portal.techmazone.com.pk/orders/new,1004,ISB


In [43]:
# There are many rows, strict pat
strict_pattern = transactions24["warehouse_code"].str.extract(r"WH-([A-Z]{3})-", expand=False).str.upper()
not_pattern = transactions24["warehouse_code"].str.extract(r"WH[-/ ]([\w+]{3})[-/ ]", expand=False).str.upper()

difference = strict_pattern != not_pattern
difference.value_counts()

print("Difference Between Strict and Not Strict Pattern")
print(pd.concat([strict_pattern[difference],
                 not_pattern[difference]], axis=1))


Difference Between Strict and Not Strict Pattern
     warehouse_code warehouse_code
1               NaN            SKT
2               NaN            NaN
3               NaN            MUL
5               NaN            MUL
6               NaN            NaN
...             ...            ...
1693            NaN            QTA
1695            NaN            PEW
1696            NaN            NaN
1697            NaN            LHR
1699            NaN            ISB

[1274 rows x 2 columns]



### Question D3  
*6 marks - Hard*

The `order_note` column contains bracketed status tags such as `[RUSH]`, `[pending]`, `[pdf]`.

Using **anchors**, report three separate counts for `transactions_2023`:

1. notes that **begin** with a tag
2. notes that **end** with a tag
3. notes that are **nothing but** a tag

Then show two example notes from each of the three groups.

In [44]:
order_note = transactions23["order_note"]
order_note

0                  [RUSH] deliver before friday
1               [URGENT] client waiting at site
2               [HOLD] awaiting credit approval
3               [PRIORITY] escalated by manager
4        [BACKORDER] awaiting stock from vendor
                         ...                   
1395            [PRIORITY] escalated by manager
1396     [BACKORDER] awaiting stock from vendor
1397               [SPLIT] shipping in two lots
1398    confirm stock before dispatch [pending]
1399    packaging damaged on arrival [reported]
Name: order_note, Length: 1400, dtype: object

In [45]:
pattern = r"^\[\w+]"
count_one = order_note[order_note.str.contains(pattern)]
count_one

0                 [RUSH] deliver before friday
1              [URGENT] client waiting at site
2              [HOLD] awaiting credit approval
3              [PRIORITY] escalated by manager
4       [BACKORDER] awaiting stock from vendor
                         ...                  
1393           [URGENT] client waiting at site
1394           [HOLD] awaiting credit approval
1395           [PRIORITY] escalated by manager
1396    [BACKORDER] awaiting stock from vendor
1397              [SPLIT] shipping in two lots
Name: order_note, Length: 354, dtype: object

In [46]:
pattern = r"\[\w+]$"
count_two = order_note[order_note.str.contains(pattern)]
count_two

6           confirm stock before dispatch [pending]
7           packaging damaged on arrival [reported]
8                  customer asked for invoice [pdf]
9        installation scheduled next week [planned]
10      credit terms extended to 60 days [approved]
                           ...                     
1359                                      [PENDING]
1360                                       [CLOSED]
1361                                      [flagged]
1398        confirm stock before dispatch [pending]
1399        packaging damaged on arrival [reported]
Name: order_note, Length: 350, dtype: object

In [47]:
pattern = r"^\[\w+]$"
count_three = order_note[order_note.str.contains(pattern)]
count_three

12       [VERIFIED]
13      [CANCELLED]
14            [pdf]
15        [PENDING]
16         [CLOSED]
           ...     
1357    [CANCELLED]
1358          [pdf]
1359      [PENDING]
1360       [CLOSED]
1361      [flagged]
Name: order_note, Length: 174, dtype: object


### Question D4  
*6 marks - Hard*

Some notes contain an accidental **repeated word**, either immediately (`urgent urgent needs dispatch`) or with one word between them (`quote sent quote by courier`).

Using **backreferences**:

1. Count and list the notes where a word repeats **immediately**.
2. Count and list the notes where a word repeats with **exactly one word in between**.
3. In a comment, explain why the pattern for part 2 must not simply be the part 1 pattern with an extra `\s\w+`.

In [48]:
pattern = r"\b(\w+)\s\1\b"
immediate_repeating_count = order_note[order_note.str.contains(pattern)]
immediate_repeating_count

C:\Users\PC\AppData\Local\Temp\ipykernel_16320\4114421612.py:2: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.
  immediate_repeating_count = order_note[order_note.str.contains(pattern)]


18      please please confirm the delivery slot
19        urgent urgent needs same day dispatch
20         call the the customer before loading
21                  send send updated quotation
22       check check the serial before handover
                         ...                   
1363      urgent urgent needs same day dispatch
1364       call the the customer before loading
1365                send send updated quotation
1366     check check the serial before handover
1367         very very late response from buyer
Name: order_note, Length: 174, dtype: object

In [49]:
pattern = r"\b(\w+)\s+\w+\s+\1\b"
repeating_count_word_in_b_w = order_note[order_note.str.contains(pattern)]
repeating_count_word_in_b_w

C:\Users\PC\AppData\Local\Temp\ipykernel_16320\3497688944.py:2: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.
  repeating_count_word_in_b_w = order_note[order_note.str.contains(pattern)]


24            quote sent quote by courier
25      invoice raised invoice against PO
26       stock checked stock at warehouse
27      order placed order through portal
28       unit tested unit before dispatch
                      ...                
1368          quote sent quote by courier
1369    invoice raised invoice against PO
1370     stock checked stock at warehouse
1371    order placed order through portal
1372     unit tested unit before dispatch
Name: order_note, Length: 145, dtype: object

In [50]:
# if we had added [\s\w+.] in ['r"\b(\w+)\s\1\b"], this would add a third word followed by same second word
pattern1 = r"\b(\w+)\s\1\b"
pattern2 = r"\b(\w+)\s+\w+\s+\1\b"


### Question D5  
*6 marks - Hard*

Two questions on `product_details`, using **lookarounds** and **word boundaries**:

1. List every product whose name contains `Book` **not preceded by** `Mac`. (`ProBook`, `ZenBook` count; `MacBook` must not.)
2. List every product containing `pro` as a **standalone word**, case-insensitive. (`MacBook Pro` counts; `ProBook` must not.)
3. In a comment, explain why part 2 cannot be solved with `str.contains("Pro")`.

In [51]:
product_details = products["product_details"]
product_details

0         Dell Latitude 5420 15.6inch
1         Dell Latitude 7430 16.0inch
2           Dell Vostro 3510 17.3inch
3           Dell Vostro 5620 13.3inch
4         Dell Inspiron 3511 14.0inch
                   ...               
59      Samsung Galaxy Book2 14.0inch
60      Samsung Galaxy Book3 15.6inch
61    Samsung Notebook Flash 16.0inch
62        Samsung Odyssey 15 17.3inch
63      Samsung Galaxy Book4 13.3inch
Name: product_details, Length: 64, dtype: object

In [52]:
pattern = r"(?<![Mm]ac)[Bb]ook"
product_details[product_details.str.lower().str.contains(pattern)]
# product_details.sample()

9                HP ProBook 450 14.0inch
10               HP ProBook 640 15.6inch
11             HP EliteBook 840 16.0inch
12             HP EliteBook 650 17.3inch
17                  HP ZBook 15 17.3inch
18          Lenovo ThinkBook 15 13.3inch
19          Lenovo ThinkBook 14 14.0inch
35          Acer Chromebook 314 15.6inch
36           Asus ExpertBook B1 16.0inch
37           Asus ExpertBook B5 17.3inch
38              Asus ZenBook 14 13.3inch
39              Asus ZenBook 15 14.0inch
40             Asus VivoBook 15 15.6inch
41             Asus VivoBook 16 16.0inch
58    Samsung Galaxy Chromebook 13.3inch
59         Samsung Galaxy Book2 14.0inch
60         Samsung Galaxy Book3 15.6inch
61       Samsung Notebook Flash 16.0inch
63         Samsung Galaxy Book4 13.3inch
Name: product_details, dtype: object

In [53]:
# List every product containing pro as a standalone word, case-insensitive. (MacBook Pro counts; ProBook must not.)
pattern =  r"\b[Pp]ro\b"
product_details[product_details.str.lower().str.contains(pattern)]

47    Apple MacBook Pro 14 14.0inch
48    Apple MacBook Pro 16 16.0inch
Name: product_details, dtype: object

- Correct_pattern = ` r"\b[Pp]ro\b"`
- incorrect_pattern = `r"Pro"`

- REASON BEHIND not using `r"Pro"` pattern is because it fails to fetch neither **MacBook Pro** OR **ProBook**


### Question D6  
*6 marks - Hard*

Two final regex tasks on `transactions_2023`:

1. From `source_url`, use **named capturing groups** to split each URL into `protocol`, `domain` and `path` in a single extraction. Report how many orders came from each domain and how many used `https`.
2. The notes spell the same word four ways: `e-mail`, `e mail`, `email`, `E-Mail`. Show the four spellings and their counts, then rewrite the column so all four become `email`. Prove it worked.

In [54]:
# 1
url = transactions23["source_url"]
url

0          https://portal.techmazone.com.pk/orders/new
1                    http://b2b.dell-partners.pk/quote
2                     https://shop.hpstore.pk/checkout
3                  http://orders.lenovo-pk.com/b2b/new
4                 https://www.acer-distributor.pk/cart
                             ...                      
1395               http://orders.lenovo-pk.com/b2b/new
1396              https://www.acer-distributor.pk/cart
1397    https://portal.techmazone.com.pk/orders/repeat
1398               http://b2b.dell-partners.pk/renewal
1399          https://store.asus-pk.com/business/order
Name: source_url, Length: 1400, dtype: object

In [55]:
# pattern = r"(?P<Protocol>https?:)
pattern = r"(?P<Protocol>https:?//)"
protocol = url.str.extract(pattern)
protocol

,Protocol
0,https://
1,NaN
2,https://
3,NaN
4,https://
...,...
1395,NaN
1396,https://
1397,https://
1398,NaN


In [56]:
pattern = r"(?P<Protocol>https?)://(?P<Domain>[\w.]+)"
domain = url.str.extract(pattern)
domain

,Protocol,Domain
0,https,portal.techmazone.com.pk
1,http,b2b.dell
2,https,shop.hpstore.pk
3,http,orders.lenovo
4,https,www.acer
...,...,...
1395,http,orders.lenovo
1396,https,www.acer
1397,https,portal.techmazone.com.pk
1398,http,b2b.dell


In [57]:
pattern = r"(?P<Protocol>https?)://(?P<Domain>[\w.]+)(?P<Path>[\w./]*)"
protocol_domain_path = url.str.extract(pattern)
protocol_domain_path

,Protocol,Domain,Path
0,https,portal.techmazone.com.pk,/orders/new
1,http,b2b.dell,
2,https,shop.hpstore.pk,/checkout
3,http,orders.lenovo,
4,https,www.acer,
...,...,...,...
1395,http,orders.lenovo,
1396,https,www.acer,
1397,https,portal.techmazone.com.pk,/orders/repeat
1398,http,b2b.dell,


In [58]:
# 817 used 'https://' protocol in their website
protocol[protocol == "https://"].value_counts()

Protocol
https://    817
Name: count, dtype: int64

domain.value_counts()

- #2 The notes spell the same word four ways`: e-mail, e mail, email, E-Mail.` Show the four spellings and their counts, then rewrite the column so all four become `email`. Prove it worked.

In [59]:
pattern = r"(e[\s-]?mail)"
order_note.str.extract(pattern,flags=re.I).value_counts()

0     
E mail    29
E-Mail    29
e mail    29
e-mail    29
email     29
Name: count, dtype: int64

In [60]:
transactions23["order_note"]

0                  [RUSH] deliver before friday
1               [URGENT] client waiting at site
2               [HOLD] awaiting credit approval
3               [PRIORITY] escalated by manager
4        [BACKORDER] awaiting stock from vendor
                         ...                   
1395            [PRIORITY] escalated by manager
1396     [BACKORDER] awaiting stock from vendor
1397               [SPLIT] shipping in two lots
1398    confirm stock before dispatch [pending]
1399    packaging damaged on arrival [reported]
Name: order_note, Length: 1400, dtype: object

In [61]:
#simply apply the above pattern using replace
pattern = r"(e[\s-]?mail)"
transactions23["order_note"]= transactions23["order_note"].str.replace(pattern,"email",regex=True, flags=re.I)
transactions23["order_note"]

0                  [RUSH] deliver before friday
1               [URGENT] client waiting at site
2               [HOLD] awaiting credit approval
3               [PRIORITY] escalated by manager
4        [BACKORDER] awaiting stock from vendor
                         ...                   
1395            [PRIORITY] escalated by manager
1396     [BACKORDER] awaiting stock from vendor
1397               [SPLIT] shipping in two lots
1398    confirm stock before dispatch [pending]
1399    packaging damaged on arrival [reported]
Name: order_note, Length: 1400, dtype: object

In [62]:
pattern = r"(e[\s-]?mail)"
transactions23["order_note"].str.extract(pattern,flags=re.I).value_counts()

0    
email    145
Name: count, dtype: int64

---
## Part E - Combining the Tables

*14 marks*

Six files, one analysis table. Nothing in this section raises an exception. It just gives you the wrong answer.



### Question E1  
*6 marks - Hard*

1. Rename the 2024 columns to match 2023, add a `year` column to each file, and stack them into one DataFrame `tx` with a clean `0..n` index.
2. Some order IDs were entered twice. Find every repeated ID and show the offending rows.

**You may not use `duplicated()`, `drop_duplicates()` or `isin()`.**

In [69]:
transactions23["Year"] = 2023
transactions24["Year"] = 2024

In [63]:
transactions24.columns

Index(['order_no', 'customer_id', 'item_code', 'quantity', 'representative',
       'warehouse_code', 'discount', 'month', 'note', 'source_link', 'cust_id',
       'region_code'],
      dtype='object')

In [64]:
transactions23.columns

Index(['order_id', 'customer_code', 'product_id', 'qty', 'sales_rep',
       'warehouse', 'discount_pct', 'order_month', 'order_note', 'source_url',
       'cust_id', 'region_code'],
      dtype='object')

In [76]:
# Using .rename() function, renaming the transaction_24 column to transaction_23 
transactions24.rename(
    {"order_no":"order_id","customer_id":"customer_code",
     "item_code":"product_id","quantity":"qty",
     "representative":"sales_rep","warehouse_code":"warehouse",
    "discount":"discount_pct","month":"order_month","note":"order_note",
    "source_link":"source_url"},
    axis=1, inplace=True
)

In [81]:
print(transactions23.columns, transactions24.columns)

Index(['order_id', 'customer_code', 'product_id', 'qty', 'sales_rep',
       'warehouse', 'discount_pct', 'order_month', 'order_note', 'source_url',
       'cust_id', 'region_code', 'Year'],
      dtype='object') Index(['order_id', 'customer_code', 'product_id', 'qty', 'sales_rep',
       'warehouse', 'discount_pct', 'order_month', 'order_note', 'source_url',
       'cust_id', 'region_code', 'Year'],
      dtype='object')


In [112]:
# Merging:
tx = pd.concat([transactions23,transactions24], ignore_index=True)
tx.head(3)

,order_id,customer_code,product_id,qty,sales_rep,warehouse,discount_pct,order_month,order_note,source_url,cust_id,region_code,Year
0,ORD23-70001,CUST-PK-1221,PRD-871,17,Omar Siddiqui,WH-HYD-11,0.0,Jun,[RUSH] deliver before friday,https://portal.techmazone.com.pk/orders/new,1221,HYD,2023
1,ORD23-70002,cust_pk_1239,PRD-864,3,Omar Siddiqui,WH/ISB/12,NaN,Jul,[URGENT] client waiting at site,http://b2b.dell-partners.pk/quote,1239,ISB,2023
2,ORD23-70003,Customer: PK-1178 (2022),PRD-864,19,Omar Siddiqui,wh-skt-12,7.5,Aug,[HOLD] awaiting credit approval,https://shop.hpstore.pk/checkout,1178,NaN,2023


In [115]:
id_count = transactions["order_id"].value_counts()
id_count

order_id
ORD24-90402    2
ORD24-90913    2
ORD24-91201    2
ORD24-90138    2
ORD24-90687    2
              ..
ORD23-71034    1
ORD23-71035    1
ORD23-71036    1
ORD23-71037    1
ORD24-91694    1
Name: count, Length: 3094, dtype: int64

In [116]:
repeating_id = id_count[id_count > 1]
repeating_id

order_id
ORD24-90402    2
ORD24-90913    2
ORD24-91201    2
ORD24-90138    2
ORD24-90687    2
ORD24-91462    2
Name: count, dtype: int64

In [124]:
offending_rows = tx[tx["order_id"].map(repeating_id)]
offending_rows

0      NaN
1      NaN
2      NaN
3      NaN
4      NaN
        ..
3095   NaN
3096   NaN
3097   NaN
3098   NaN
3099   NaN
Name: order_id, Length: 3100, dtype: float64


### Question E2  
*8 marks - Hard*

1. Join `tx` to `products`, `customers` and `regions` into one table `m`. State in a comment which join type you used and why, and prove no rows were gained or lost.
2. Which **registered customers have never placed an order**?
3. Which **catalogue products were never sold**?

**You may not use `isin()`, `isnull()` or `notnull()`.** Build the membership test yourself.

In [151]:
regions

,region_code,region_name,city,country
0,KHI,South,Karachi,Pakistan
1,HYD,South,Hyderabad,Pakistan
2,LHR,Central,Lahore,Pakistan
3,FSD,Central,Faisalabad,Pakistan
4,MUL,Central,Multan,Pakistan
5,SKT,Central,Sialkot,Pakistan
6,ISB,North,Islamabad,Pakistan
7,RWP,North,Rawalpindi,Pakistan
8,PEW,North West,Peshawar,Pakistan
9,QTA,West,Quetta,Pakistan


In [153]:
# Here we used, 'left' join bcz tx was the master table
# Using 'inner' join would drop missing values
m = tx.merge(products,on='product_id', how="left").merge(customers, on='cust_id', how="left").merge(regions, on='region_code', how="left")
m

,order_id,customer_code_x,product_id,qty,sales_rep,warehouse,discount_pct,order_month,order_note,source_url,...,brand,screen_size_inches,customer_code_y,customer_name,segment_type,home_city,signup_year,region_name,city,country
0,ORD23-70001,CUST-PK-1221,PRD-871,17,Omar Siddiqui,WH-HYD-11,0.0,Jun,[RUSH] deliver before friday,https://portal.techmazone.com.pk/orders/new,...,MSI,17.3inch,CUST-PK-1221,Chand Suppliers,Corporate,Karachi,2015,South,Hyderabad,Pakistan
1,ORD23-70002,cust_pk_1239,PRD-864,3,Omar Siddiqui,WH/ISB/12,NaN,Jul,[URGENT] client waiting at site,http://b2b.dell-partners.pk/quote,...,MSI,16.0inch,cust_pk_1239,Tariq Suppliers,Corporate,Peshawar,2023,North,Islamabad,Pakistan
2,ORD23-70003,Customer: PK-1178 (2022),PRD-864,19,Omar Siddiqui,wh-skt-12,7.5,Aug,[HOLD] awaiting credit approval,https://shop.hpstore.pk/checkout,...,MSI,16.0inch,Customer: PK-1178 (2022),Qasim Systems,Reseller,Rawalpindi,2022,NaN,NaN,NaN
3,ORD23-70004,2019-PK_1035,PRD-885,12,Omar Siddiqui,WH SKT 08,10.0,Sep,[PRIORITY] escalated by manager,http://orders.lenovo-pk.com/b2b/new,...,MSI,14.0inch,2019-PK_1035,Shazia Enterprises,Reseller,Multan,2019,Central,Sialkot,Pakistan
4,ORD23-70005,CUST-PK-1216,PRD-794,14,Omar Siddiqui,WH-FSD-03,7.5,Oct,[BACKORDER] awaiting stock from vendor,https://www.acer-distributor.pk/cart,...,Asus,16.0inch,CUST-PK-1216,Fahad Enterprises,Reseller,Sialkot,2020,Central,Faisalabad,Pakistan
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3095,ORD24-91690,BR2016/PK-1072,PRD-528,7,Hamza Sheikh,WH/PEW/07,5.0,Nov,installation scheduled next week [planned],https://www.samsung-b2b.pk/orders/create,...,Dell,13.3inch,BR2016/PK-1072,Hina Traders,Reseller,Hyderabad,2016,North West,Peshawar,Pakistan
3096,ORD24-91691,Customer: PK-1078 (2022),PRD-738,40,Hamza Sheikh,wh-qta-03,10.0,Dec,credit terms extended to 60 days [approved],https://shop.hpstore.pk/enterprise,...,Acer,13.3inch,Customer: PK-1078 (2022),Lubna Corporation,Corporate,Rawalpindi,2022,NaN,NaN,NaN
3097,ORD24-91692,2019-PK_1005,PRD-626,12,Hamza Sheikh,WH LHR 12,5.0,Jan,serial numbers shared with buyer [done],http://orders.lenovo-pk.com/renewal,...,HP,17.3inch,2019-PK_1005,Fahad Traders,Corporate,Multan,2019,Central,Lahore,Pakistan
3098,ORD24-91693,cust_pk_1004,PRD-668,7,Hamza Sheikh,WH-ISB-03,7.5,Feb,[VERIFIED],https://portal.techmazone.com.pk/orders/new,...,Lenovo,13.3inch,cust_pk_1004,Yusra Electronics,Retail,Faisalabad,2018,North,Islamabad,Pakistan


In [156]:
# Verifying
print(f"tx rows:",len(tx))
print(f"m rows:",len(m))
print(f"Difference of rows b/w them:",len(m)-len(tx))

#This confirms there are no rows that are either gained or lost during this merging

tx rows: 3100
m rows: 3100
Difference of rows b/w them: 0


In [174]:
#2. Which registered customers have never placed an order?

# these are customers that have placed order, extracting them using `unique()` function and customer_id column
reg_customer = m["cust_id"].unique()
reg_customer

array([1221, 1239, 1178, 1035, 1216, 1159, 1208, 1084, 1135, 1123, 1132,
       1007, 1172, 1190, 1157, 1166, 1043, 1077, 1198, 1130, 1014, 1117,
       1079, 1085, 1137, 1232, 1032, 1075, 1030, 1223, 1017, 1099, 1143,
       1027, 1155, 1045, 1119, 1121, 1083, 1201, 1024, 1156, 1070, 1204,
       1207, 1022, 1185, 1170, 1064, 1181, 1212, 1184, 1006, 1225, 1012,
       1068, 1106, 1034, 1211, 1197, 1219, 1141, 1050, 1150, 1183, 1236,
       1009, 1237, 1040, 1229, 1147, 1025, 1228, 1125, 1028, 1144, 1013,
       1101, 1171, 1138, 1031, 1158, 1194, 1053, 1133, 1029, 1193, 1088,
       1118, 1114, 1196, 1069, 1187, 1189, 1163, 1089, 1139, 1063, 1109,
       1023, 1080, 1148, 1112, 1078, 1073, 1052, 1103, 1124, 1042, 1233,
       1057, 1048, 1176, 1108, 1127, 1186, 1018, 1054, 1210, 1142, 1238,
       1039, 1162, 1160, 1177, 1140, 1226, 1214, 1061, 1104, 1003, 1010,
       1049, 1097, 1206, 1227, 1180, 1015, 1199, 1066, 1220, 1149, 1095,
       1205, 1102, 1105, 1016, 1019, 1072, 1059, 12

In [180]:
customer_no_order_placed = customers[customers["cust_id"].isin(reg_customer)]
customer_no_order_placed

print(f"Number of total customers never ordered:",len(customer_no_order_placed))

Number of total customers never ordered: 240


In [187]:
m.head(3)

,order_id,customer_code_x,product_id,qty,sales_rep,warehouse,discount_pct,order_month,order_note,source_url,...,brand,screen_size_inches,customer_code_y,customer_name,segment_type,home_city,signup_year,region_name,city,country
0,ORD23-70001,CUST-PK-1221,PRD-871,17,Omar Siddiqui,WH-HYD-11,0.0,Jun,[RUSH] deliver before friday,https://portal.techmazone.com.pk/orders/new,...,MSI,17.3inch,CUST-PK-1221,Chand Suppliers,Corporate,Karachi,2015,South,Hyderabad,Pakistan
1,ORD23-70002,cust_pk_1239,PRD-864,3,Omar Siddiqui,WH/ISB/12,NaN,Jul,[URGENT] client waiting at site,http://b2b.dell-partners.pk/quote,...,MSI,16.0inch,cust_pk_1239,Tariq Suppliers,Corporate,Peshawar,2023,North,Islamabad,Pakistan
2,ORD23-70003,Customer: PK-1178 (2022),PRD-864,19,Omar Siddiqui,wh-skt-12,7.5,Aug,[HOLD] awaiting credit approval,https://shop.hpstore.pk/checkout,...,MSI,16.0inch,Customer: PK-1178 (2022),Qasim Systems,Reseller,Rawalpindi,2022,NaN,NaN,NaN


In [188]:
#3. Which catalogue products were never sold?

# these are products that are sold, extracting using `unique()` function and `product_id` column
products_sold = m["product_id"].unique()
products_sold

array(['PRD-871', 'PRD-864', 'PRD-885', 'PRD-794', 'PRD-927', 'PRD-759',
       'PRD-521', 'PRD-675', 'PRD-766', 'PRD-773', 'PRD-584', 'PRD-570',
       'PRD-661', 'PRD-556', 'PRD-605', 'PRD-941', 'PRD-542', 'PRD-822',
       'PRD-843', 'PRD-633', 'PRD-647', 'PRD-577', 'PRD-507', 'PRD-654',
       'PRD-857', 'PRD-703', 'PRD-612', 'PRD-878', 'PRD-836', 'PRD-787',
       'PRD-808', 'PRD-682', 'PRD-598', 'PRD-850', 'PRD-710', 'PRD-626',
       'PRD-668', 'PRD-619', 'PRD-752', 'PRD-913', 'PRD-731', 'PRD-899',
       'PRD-745', 'PRD-563', 'PRD-591', 'PRD-528', 'PRD-801', 'PRD-738',
       'PRD-717', 'PRD-514', 'PRD-689', 'PRD-780', 'PRD-948', 'PRD-892',
       'PRD-535', 'PRD-696', 'PRD-829', 'PRD-920', 'PRD-934'],
      dtype=object)

In [190]:
products_unsold = products[products["product_id"].isin(products_sold)]
products_unsold

print(f"Number of total products never sold:",len(products_unsold))

Number of total products never sold: 59


---
## Part F - Aggregation and Analysis

*11 marks*

Answer the questions the business actually asked.



### Question F1  
*5 marks - Medium*

1. Add `gross_revenue` (quantity x unit price) and `net_revenue` (after discount) to `m`. Report the total net revenue and the total discount given away.
2. Using `groupby`, report the number of orders and the total net revenue per **sales rep**.
3. Use `get_group` to show the first three transactions of any one rep.

In [201]:
m["gross_revenue"] = m["qty"] * m["unit_price"]
m.head(3)

,order_id,customer_code_x,product_id,qty,sales_rep,warehouse,discount_pct,order_month,order_note,source_url,...,screen_size_inches,customer_code_y,customer_name,segment_type,home_city,signup_year,region_name,city,country,gross_revenue
0,ORD23-70001,CUST-PK-1221,PRD-871,17,Omar Siddiqui,WH-HYD-11,0.0,Jun,[RUSH] deliver before friday,https://portal.techmazone.com.pk/orders/new,...,17.3inch,CUST-PK-1221,Chand Suppliers,Corporate,Karachi,2015,South,Hyderabad,Pakistan,2286500
1,ORD23-70002,cust_pk_1239,PRD-864,3,Omar Siddiqui,WH/ISB/12,NaN,Jul,[URGENT] client waiting at site,http://b2b.dell-partners.pk/quote,...,16.0inch,cust_pk_1239,Tariq Suppliers,Corporate,Peshawar,2023,North,Islamabad,Pakistan,336000
2,ORD23-70003,Customer: PK-1178 (2022),PRD-864,19,Omar Siddiqui,wh-skt-12,7.5,Aug,[HOLD] awaiting credit approval,https://shop.hpstore.pk/checkout,...,16.0inch,Customer: PK-1178 (2022),Qasim Systems,Reseller,Rawalpindi,2022,NaN,NaN,NaN,2128000


In [205]:
m["net_revenue"] = (m["qty"] * m["unit_price"]) - m["discount_pct"]
m.head(3)

,order_id,customer_code_x,product_id,qty,sales_rep,warehouse,discount_pct,order_month,order_note,source_url,...,customer_code_y,customer_name,segment_type,home_city,signup_year,region_name,city,country,gross_revenue,net_revenue
0,ORD23-70001,CUST-PK-1221,PRD-871,17,Omar Siddiqui,WH-HYD-11,0.0,Jun,[RUSH] deliver before friday,https://portal.techmazone.com.pk/orders/new,...,CUST-PK-1221,Chand Suppliers,Corporate,Karachi,2015,South,Hyderabad,Pakistan,2286500,2286500.0
1,ORD23-70002,cust_pk_1239,PRD-864,3,Omar Siddiqui,WH/ISB/12,NaN,Jul,[URGENT] client waiting at site,http://b2b.dell-partners.pk/quote,...,cust_pk_1239,Tariq Suppliers,Corporate,Peshawar,2023,North,Islamabad,Pakistan,336000,NaN
2,ORD23-70003,Customer: PK-1178 (2022),PRD-864,19,Omar Siddiqui,wh-skt-12,7.5,Aug,[HOLD] awaiting credit approval,https://shop.hpstore.pk/checkout,...,Customer: PK-1178 (2022),Qasim Systems,Reseller,Rawalpindi,2022,NaN,NaN,NaN,2128000,2127992.5


In [195]:
m["unit_price"]

,order_id,customer_code_x,product_id,qty,sales_rep,warehouse,discount_pct,order_month,order_note,source_url,...,brand,screen_size_inches,customer_code_y,customer_name,segment_type,home_city,signup_year,region_name,city,country
0,ORD23-70001,CUST-PK-1221,PRD-871,17,Omar Siddiqui,WH-HYD-11,0.0,Jun,[RUSH] deliver before friday,https://portal.techmazone.com.pk/orders/new,...,MSI,17.3inch,CUST-PK-1221,Chand Suppliers,Corporate,Karachi,2015,South,Hyderabad,Pakistan
1,ORD23-70002,cust_pk_1239,PRD-864,3,Omar Siddiqui,WH/ISB/12,NaN,Jul,[URGENT] client waiting at site,http://b2b.dell-partners.pk/quote,...,MSI,16.0inch,cust_pk_1239,Tariq Suppliers,Corporate,Peshawar,2023,North,Islamabad,Pakistan
2,ORD23-70003,Customer: PK-1178 (2022),PRD-864,19,Omar Siddiqui,wh-skt-12,7.5,Aug,[HOLD] awaiting credit approval,https://shop.hpstore.pk/checkout,...,MSI,16.0inch,Customer: PK-1178 (2022),Qasim Systems,Reseller,Rawalpindi,2022,NaN,NaN,NaN



### Question F2  
*6 marks - Hard*

1. Build a pivot table with **region name** down the side, **category** across the top, and **total** net revenue in the cells.
2. For each region, find the single **brand** that earned the most net revenue, and how much.

For part 2 you must use a **loop over the groups** and track the maximum yourself - do not use `idxmax`, `max`, `nlargest` or `sort_values`.